hazelcast/src/main/java/com/hazelcast/internal/networking/nio/NioOutboundPipeline.java

```diff
@@ -226,6 +218,25 @@ public void write(OutboundFrame frame) {
         }
     }
 
+    // executes the pipeline. Either on the calling thread or on th owning NIO thread.
+    private void executePipeline() {
+         if (writeThroughEnabled && !concurrencyDetection.isDetected()) {
+            // we are allowed to do a write through, so lets process the request on the calling thread
+            try {
+                process();
+            } catch (Throwable t) {
+                onError(t);
+            }
+        } else {
+            if (selectionKeyWakeupEnabled) {
+                registerOp(OP_WRITE);
+                selectionKey.selector().wakeup();
```

Comment:
I wonder if this and the previous lines are safe. `selectionKey` could be changed to `null` or the new key from NIO thread on migration. So in principle, we may observe three values: old key, new key or `null`. Observing an old key doesn't seem to be of great concern since initial selection keys are re-initialized on migration, and `OP_WRITE` will be registered anyway. But `null` is a problem unless I miss some implicit synchronization. 
Ideally, we'd better merge `registerOp` and wakeup to a single method, which will do a null-check and ensure that three calls to a `selectionKey` in these two lines of code are performed on the same instance of the key.

Reply:
Before a migration, the scheduled flag is set and it will be unset when the migration completes. So it can't happen that during a migration a different thread will schedule the pipeline.

Reply:
Discussed with Peter. Safety is guaranteed by the fact that the pipeline starts in a `SCHEDULED` state, and migration is only possible in this state. So it is not possible that one thread will register a write key, while the other is doing a migration.